File: Python/Medium/3397_Maximum_Number_of_Distinct_Elements_After_Operations/solution.py
class Solution:
    def maxDistinctElements(self, nums: list[int], k: int) -> int:
        nums.sort()
        prev = float("-inf")
        count = 0

        for i in range(len(nums)):
            current = min(max(nums[i] - k, prev + 1), nums[i] + k)
            if current > prev:
                count += 1
                prev = current 

        return count

File: Golang/Medium/3397_Maximum_Number_of_Distinct_Elements_After_Operations/solution.go
package maximumnumberofdistinctelementsafteroperations

import (
	"math"
	"sort"
)

func maxDistinctElements(nums []int, k int) int {
	sort.Ints(nums)
	prev := math.MinInt
	count := 0

	for _, num := range nums {
		current := min(max(num-k, prev+1), num+k)
		if current > prev {
			count += 1
			prev = current
		}
	}
	return count
}
